keras/preprocessing/image.py

```diff
@@ -643,16 +651,19 @@ def __init__(self, directory, image_data_generator,
         i = 0
         for subdir in classes:
             subpath = os.path.join(directory, subdir)
-            for fname in sorted(os.listdir(subpath)):
-                is_valid = False
-                for extension in white_list_formats:
-                    if fname.lower().endswith('.' + extension):
-                        is_valid = True
-                        break
-                if is_valid:
-                    self.classes[i] = self.class_indices[subdir]
-                    self.filenames.append(os.path.join(subdir, fname))
-                    i += 1
+            subpath_len = len(directory) + 1  # assuming separator is 1 character
+            for root, dirs, files in _recursive_list(subpath):
+                for fname in files:
+                    is_valid = False
+                    for extension in white_list_formats:
+                        if fname.lower().endswith('.' + extension):
+                            is_valid = True
+                            break
+                    if is_valid:
+                        self.classes[i] = self.class_indices[subdir]
+                        # add filename relative to subpath
+                        self.filenames.append(os.path.join(root, fname)[subpath_len:])
```

Comment:
Hmmm... not really. If `directory` is an absolute path `/data/cats-vs-dogs/train/`, the file under `/data/cats-vs-dogs/train/cats/photos-of-wild-cats/lion.jpg` should appear in `self.filenames` relative to `directory`, so `cats/photos-of-wild-cats/lion.jpg`

That being said, I'm changing it to use `os.path.relpath` instead of string operations.